x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -25,21 +33,150 @@ import {
+    }
+  };
+
+  /**
+   * Gets all {@link SessionSavedObjectAttributes | Background Searches} that
+   * currently being tracked by the service.
+   *	   *
+   * @remarks
+   * Uses `internalSavedObjectsClient` as this is called asynchronously, not within the
+   * context of a user's session.
+   */
+  private async getAllMappedSavedObjects() {
+    const activeMappingIds = Array.from(this.sessionSearchMap.keys())
+      .map((sessionId) => `"${sessionId}"`)
+      .join(' | ');
+    const res = await this.internalSavedObjectsClient.find<BackgroundSessionSavedObjectAttributes>({
+      perPage: INMEM_MAX_SESSIONS, // If there are more sessions in memory, they will be synced when some items are cleared out.
+      type: BACKGROUND_SESSION_TYPE,
+      search: activeMappingIds,
+      searchFields: ['sessionId'],
+      namespaces: ['*'],
+    });
+    this.logger.debug(`getAllMappedSavedObjects | Got ${res.saved_objects.length} items`);
+    return res.saved_objects;
+  }
+
+  private clearSessions = () => {
+    const curTime = moment();
+
+    this.sessionSearchMap.forEach((sessionInfo, sessionId) => {
+      if (
+        moment.duration(curTime.diff(sessionInfo.insertTime)).asSeconds() >
+        INMEM_TRACKING_TIMEOUT_SEC
+      ) {
+        this.logger.debug(`clearSessions | Deleting expired session ${sessionId}`);
+        this.sessionSearchMap.delete(sessionId);
+      } else if (sessionInfo.retryCount >= MAX_UPDATE_RETRIES) {
+        this.logger.warn(`clearSessions | Deleting failed session ${sessionId}`);
+        this.sessionSearchMap.delete(sessionId);
+      }
+    });
+  };
+
+  public async monitorMappedIds() {
+    try {
+      this.clearSessions();
+
+      if (!this.sessionSearchMap.size) return;
+      this.logger.debug(`monitorMappedIds | Map contains ${this.sessionSearchMap.size} items`);
+
+      const savedSessions = await this.getAllMappedSavedObjects();
+      const updatedSessions = await this.updateAllSavedObjects(savedSessions);
+
+      updatedSessions.forEach((updatedSavedObject) => {
+        const sessionInfo = this.sessionSearchMap.get(updatedSavedObject.id)!;
+        if (updatedSavedObject.error) {
+          // Retry next time
+          sessionInfo.retryCount++;
+        } else if (updatedSavedObject.attributes.idMapping) {
+          // Delete the ids that we just saved, avoiding a potential new ids being lost.
+          Object.keys(updatedSavedObject.attributes.idMapping).forEach((key) => {
+            sessionInfo.ids.delete(key);
+          });
+          // If the session object is empty, delete it as well
+          if (!sessionInfo.ids.entries.length) {
+            this.sessionSearchMap.delete(updatedSavedObject.id);
+          } else {
+            sessionInfo.retryCount = 0;
+          }
+        }
+      });
+    } catch (e) {
+      this.logger.error(`monitorMappedIds | Error fetching sessions. ${e}`);
```

Comment:
wouldn't `Error fetching sessions.` be misleading in case error happen not because of `await this.getAllMappedSavedObjects();` but for example because of `await this.updateAllSavedObjects(savedSessions);` ?